packages/flutter_tools/lib/src/ios/ios_deploy.dart

```diff
@@ -598,29 +598,66 @@ class IOSDeployDebugger {
     if (!debuggerAttached) {
       return;
     }
-    try {
-      // Stop the app, which will prompt the backtrace to be printed for all threads in the stdoutSubscription handler.
-      _iosDeployProcess?.stdin.writeln(_signalStop);
-    } on SocketException catch (error) {
-      // Best effort, try to detach, but maybe the app already exited or already detached.
-      _logger.printTrace('Could not stop app from debugger: $error');
-    }
+    // Stop the app, which will prompt the backtrace to be printed for all
+    // threads in the stdoutSubscription handler.
+    await stdinWriteln(
+      _signalStop,
+      onError: (Object error, _) {
+        _logger.printTrace('Could not stop the app: $error');
+      },
+    );
+
     // Wait for logging to finish on process exit.
     return logLines.drain();
   }
 
-  void detach() {
-    if (!debuggerAttached) {
+  Future<void>? _stdinWriteFuture;
+
+  /// Queue write of [line] to STDIN of [_iosDeployProcess].
+  ///
+  /// No-op if [_iosDeployProcess] is null.
+  ///
+  /// This write will not happen until the flush of any previous writes have
+  /// completed, because calling [IOSink.flush()] before a previous flush has
+  /// completed will throw a [StateError].
+  ///
+  /// This method needs to keep track of the [_stdinWriteFuture] from previous
+  /// calls because the future returned by [detach] is not always await-ed.
+  Future<void> stdinWriteln(String line, {required void Function(Object, StackTrace) onError}) async {
```

Comment:
Will the next place `writelnToStdinGuarded` is adopted will also need this logic?  At which time this helper would move into `process.dart`?  Or is this logic special to this `detach`?

Reply:
I'm not a `Zone` wizard or even novice, but it seems like scheduling events in a queue so that `writelnToStdinGuarded` isn't called until the flush happens is exactly what a zone is good for?  However, my brain hurts.

Reply:
> I'm not a Zone wizard or even novice, but it seems like scheduling events in a queue so that writelnToStdinGuarded isn't called until the flush happens is exactly what a zone is good for?

Are you making a specific recommendation here? If so, can you ELI5? :)

Reply:
followed up offline, we explored different options but couldn't come up with something that worked.